x-pack/metricbeat/module/mssql/log/log.go

```diff
@@ -0,0 +1,223 @@
+
+	"github.com/elastic/beats/libbeat/common/cfgwarn"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/metricbeat/mb"
+	"github.com/elastic/beats/x-pack/metricbeat/module/mssql"
+)
+
+type dbInfo struct {
+	id   int
+	name string
+}
+
+type logSpace struct {
+	id                             int
+	totalLogSizeInBytes            int
+	usedLogSpaceInBytes            int
+	usedLogSpaceInPercent          float64
+	logSpaceInBytesSinceLastBackup int
+}
+
+type logStats struct {
+	databaseID            int
+	sizeMB                float64
+	activeSizeMB          float64
+	backupTime            string
+	sinceLastBackupMB     *float64
+	sinceLastCheckpointMB float64
+	recoverySizeMB        float64
+}
+
+func init() {
+	mb.Registry.MustAddMetricSet("mssql", "log", New,
+		mb.DefaultMetricSet(),
+		mb.WithHostParser(mssql.HostParser))
+}
+
+// MetricSet type defines all fields of the MetricSet
+type MetricSet struct {
+	mb.BaseMetricSet
+	log *logp.Logger
+	db  *sql.DB
+}
+
+// New create a new instance of the MetricSet
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	cfgwarn.Beta("The mssql log metricset is beta.")
+
+	logger := logp.NewLogger("mssql.log").With("host", base.HostData().SanitizedURI)
+
+	db, err := mssql.NewConnection(base.HostData().URI)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not create connection to db")
+	}
+
+	return &MetricSet{
+		BaseMetricSet: base,
+		log:           logger,
+		db:            db,
+	}, nil
+}
+
+// Fetch methods implements the data gathering and data conversion to the right format
+// It returns the event which is then forward to the output. In case of an error, a
+// descriptive error must be returned.
+func (m *MetricSet) Fetch(reporter mb.ReporterV2) {
+	dbs, err := m.getDbsNames()
+	if err != nil {
+		reporter.Error(err)
+		return
+	}
+
+	for _, db := range dbs {
+		moduleFields := common.MapStr{
+			"database": common.MapStr{
+				"id":   db.id,
+				"name": db.name,
+			},
+		}
+		metricsetFields := common.MapStr{}
+
+		spaceUsage, err := m.getLogSpaceUsageForDb(db.name)
+		if err != nil {
+			reporter.Error(err)
+		} else {
+			metricsetFields["space_usage"] = spaceUsage
+		}
+
+		stats, err := m.getLogStats(db)
+		if err != nil {
+			reporter.Error(err)
+		} else {
+			metricsetFields["stats"] = stats
+		}
+
+		// Both log space and log size are available, so report both
+		if isReported := reporter.Event(mb.Event{
+			ModuleFields:    moduleFields,
+			MetricSetFields: metricsetFields,
+		}); !isReported {
+			m.log.Debug("event not reported")
+		}
+
+	}
+}
+
+// Close the connection to the server at the engine level
+func (m *MetricSet) Close() error {
+	return m.db.Close()
+}
+
+func (m *MetricSet) getLogSpaceUsageForDb(dbName string) (result common.MapStr, err error) {
+
+	rows, err := m.db.Query(fmt.Sprintf(`USE %s; SELECT * FROM sys.dm_db_log_space_usage;`, dbName))
+	if err != nil {
+		return
+	}
+	defer closeRows(rows)
+
+	for rows.Next() {
```

Comment:
Uhmmm... not a bad idea, but I'll still have to leave a message for our future us so that they don't think that it's a mistake :smile:

Reply:
Well, the comment maybe needed in any case because I already though that there were some mistake here :smiley: And also Nicolas with the return. I think that without a loop it'd be more clear that only a row is expected.

Reply:
++

Reply:
Thanks for the suggestion @jsoriano !